src/Compiler/Utilities/illib.fs

```diff
@@ -1041,48 +1041,53 @@ module CancellableAutoOpens =
 
 /// Generates unique stamps
 type UniqueStampGenerator<'T when 'T: equality>() =
-    let gate = obj ()
-    let encodeTab = ConcurrentDictionary<'T, int>(HashIdentity.Structural)
-    let mutable nItems = 0
+    let encodeTab = ConcurrentDictionary<'T, Lazy<int>>(HashIdentity.Structural)
+    let mutable nItems = -1
 
-    let encode str =
-        match encodeTab.TryGetValue str with
-        | true, idx -> idx
-        | _ ->
-            lock gate (fun () ->
-                let idx = nItems
-                encodeTab[str] <- idx
-                nItems <- nItems + 1
-                idx)
+    let computeFunc = Func<'T, _>(fun _ -> lazy (Interlocked.Increment(&nItems)))
 
-    member _.Encode str = encode str
+    member _.Encode str =
+        encodeTab.GetOrAdd(str, computeFunc).Value
 
     member _.Table = encodeTab.Keys
 
 /// memoize tables (all entries cached, never collected)
 type MemoizationTable<'T, 'U>(compute: 'T -> 'U, keyComparer: IEqualityComparer<'T>, ?canMemoize) =
 
-    let table = new ConcurrentDictionary<'T, 'U>(keyComparer)
+    let table = new ConcurrentDictionary<'T, Lazy<'U>>(keyComparer)
+    let computeFunc = Func<_, _>(fun key -> lazy (compute key))
 
     member t.Apply x =
         if
             (match canMemoize with
              | None -> true
              | Some f -> f x)
         then
-            match table.TryGetValue x with
-            | true, res -> res
-            | _ ->
-                lock table (fun () ->
-                    match table.TryGetValue x with
-                    | true, res -> res
-                    | _ ->
-                        let res = compute x
-                        table[x] <- res
-                        res)
+            table.GetOrAdd(x, computeFunc).Value
         else
             compute x
 
+type internal StampedDictionairy<'T, 'U>(keyComparer: IEqualityComparer<'T>) =
+    let table = new ConcurrentDictionary<'T, Lazy<int * 'U>>(keyComparer)
+    let mutable count = -1
+
+    member _.Add(key, value) =
+        let entry = table.GetOrAdd(key, lazy (Interlocked.Increment(&count), value))
```

Comment:
Also, as much as I am not into comments in the code, here they would be really helpful actually. Why would one use this structure? How to use it then? When not to use it?

At least in my dev bubble this is not considered as mainstream coding, so explaining the intentions might help some folks looking into this file in the future.

Reply:
Thanks!